### ISSUE TITLE
Puma returns exit code 1 in container when received SIGTERM

### ISSUE THREAD

**Issue by atitan** [puma/puma#1673]
We're aware that GKE complains about puma returning exit code 1 when received SIGTERM.

Although I've been searching through previous issues but seems they're unrelated to this.

The strange thing is, it doesn't occur if I use a bash, even in the container, to execute it.

### Steps to reproduce

```
docker run {image id} bundle exec puma -C config/puma.rb
# or
docker run {image id} /bin/bash -c "bundle exec puma -C config/puma.rb"

docker kill --signal=TERM {container id}

docker container ls -a
=> Exited (1) 1 minutes ago
```

### Steps not affected
```
# on my Mac
bundle exec puma -C config/puma.rb
kill -TERM {pid}

[91324] Use Ctrl-C to stop
[91324] - Worker 0 (pid: 91358) booted, phase: 0
[91324] - Worker 1 (pid: 91359) booted, phase: 0
[91324] - Gracefully shutting down workers...
Terminated: 15
MacBook app $ echo $?
143
```

```
# inside running container's bash
docker exec -it {container id} /bin/bash
bundle exec puma -C config/puma.rb
kill -TERM {pid}

[42] Use Ctrl-C to stop
[42] - Worker 0 (pid: 57) booted, phase: 0
[42] - Worker 1 (pid: 72) booted, phase: 0
[42] - Gracefully shutting down workers...
Terminated
root@c7b7ec95ee83:/app# echo $?
143
```

### Expected behavior

Return code 15/143

### Actual behavior

Return code 1

### System configuration

**Ruby version**: 2.5.3
**Rails version**: 5.2.1
**Puma version**: 3.12.0
**Bundle version**: 1.16.6

**Comment by atitan** [comment]
Using `dumb-init` as container entrypoint solves this issue.

Although GKE is still not happy about the return code.😢

**Comment by tagliala** [comment]
Hi @atitan, I've run into this issue. Do you have any other suggestion?

The problem I'm having is that when I restart docker, successfully stopped puma containers with `on-failure` restart policy automatically restarts because their exit code is different from 0

**Comment by atitan** [comment]
Hi @tagliala, we're currently using dumb-init as init process, and set it to rewrite SIGTERM to SIGINT when forwarding signal to puma.

Hope it helps.

**Comment by fey** [comment]
For anyone who lands here: the exit status 1 in containers comes from Puma running as PID 1 (the kernel drops the SIGTERM that Ruby re-sends to itself). Details and a proposed fix are in #3941, and the discussion continues in #3639. Feedback is welcome.
